### PR TITLE
Bug 1751524 - [MYSQL8] The groups table in Bugzilla conflicts with reserved keyword in MySQL 8.x

### PR DESCRIPTION
**Pull request by dklawren**

### CODE REVIEW

File: cpanfile.snapshot [hunk 1]
```diff
@@ -22,83 +22,83 @@ DISTRIBUTIONS
       Algorithm::Diff::_impl 1.201
     requirements:
       ExtUtils::MakeMaker 0
-  Alien-Build-2.41
```
Comment (mitchhentges) [inline cpanfile.snapshot]: Is this file equivalent to a `cargo.lock`/`package-lock.json`/`requirements.txt`?
If so, would it be possible to split changing dependency versions to a different patch?

File: Bugzilla/Bug.pm [hunk 2]
```diff
@@ -4180,7 +4180,7 @@ sub groups {
       . " THEN 1 ELSE 0 END,"
       . " CASE WHEN groups.id IN($grouplist) THEN 1 ELSE 0 END,"
       . " isactive, membercontrol, othercontrol"
-      . " FROM groups"
+      . " FROM " . $dbh->quote_identifier('groups')
```
Comment (mitchhentges) [inline Bugzilla/Bug.pm:4183]: IMO, if we're already pretty MYSQL-specific, it might be simpler and easier to read to manually quote this eh?
```
...
. ' FROM "groups"'
...
```

I'm guessing that `dbh->quote_identifier()` is still a good idea if the table name comes from a variable, though.

What do you think?

File: Bugzilla/DB/Mysql.pm [hunk 3]
```diff
@@ -330,15 +330,6 @@ sub bz_setup_database {
...
 
   if ($self->utf8_charset eq 'utf8mb4') {
-    my %global = map {@$_}
-      @{$self->selectall_arrayref(q(SHOW GLOBAL VARIABLES LIKE 'innodb_%'))};
-    my $utf8mb4_supported
-      = $global{innodb_file_format} eq 'Barracuda'
-      && $global{innodb_file_per_table} eq 'ON'
-      && $global{innodb_large_prefix} eq 'ON';
-
-    die install_string('mysql_innodb_settings') unless $utf8mb4_supported;
```
Comment (mitchhentges) [inline Bugzilla/DB/Mysql.pm:340]: In the bug/commit message/PR description, can you elaborate why this is no longer needed? Perhaps link to the specific part of the release notes that obsolete this?

Comment (mitchhentges) [inline Bugzilla/DB/Mysql.pm:340]: In the bug/commit message/PR description, can you elaborate why this is no longer needed? Perhaps link to the specific part of the release notes that obsolete this?

File: Bugzilla/DB/Schema.pm [hunk 4]
```diff
@@ -2480,7 +2490,7 @@ sub _get_create_index_ddl {
   my $sql = "CREATE ";
   $sql .= "$index_type " if ($index_type && $index_type eq 'UNIQUE');
   $sql
-    .= "INDEX $index_name ON $table_name \(" . join(", ", @$index_fields) . "\)";
+    .= "INDEX \`$index_name\` ON \`$table_name\` \(" . join(", ", @$index_fields) . "\)";
```
Comment (mitchhentges) [inline Bugzilla/DB/Schema.pm:2493]: Why do the backticks need to be `\escaped` here?

File: Bugzilla/Report/Ping/Simple.pm [hunk 5]
```diff
@@ -49,8 +49,7 @@ sub _build_validator {
     creation_ts      => joi->string->required,
   });
 
-  return JSON::Validator->new(
-    schema => Mojo::JSON::Pointer->new($schema->compile));
+  return JSON::Validator->new(schema => $schema);
```
Comment (mitchhentges) [inline Bugzilla/Report/Ping/Simple.pm:52]: This is an unrelated change (probably due to the `cpan.snapshot` update) and should be in another patch IMO

File: extensions/Voting/Extension.pm [hunk 6]
```diff
@@ -899,7 +899,8 @@ sub _remove_votes {
 sub _confirm_if_vote_confirmed {
   my ($id, $product, $timestamp) = @_;
   my $bug = ref $id ? $id : new Bugzilla::Bug({id => $id, cache => 1});
-  $product ||= $bug->product_obj;
+  $product   ||= $bug->product_obj;
+  $timestamp ||= Bugzilla->dbh->selectrow_array('SELECT LOCALTIMESTAMP(0)');
```
Comment (mitchhentges) [inline extensions/Voting/Extension.pm:903]: I can't find where `$timestamp` is used 🤔

**Review by mitchhentges — CHANGES_REQUESTED** [review]
Something went funky with my github review comments, sorry. I think github choked 😨 

Anyways:
* Hopefully we can always escape table names with "`", that would be a lot more convenient for readability and review
* I'm not going to check over syntax and string-join correctness here and leave that to your hand-testing, because that will be a more effective way of avoiding regressions IMHO
* Nice work!

File: extensions/Voting/Extension.pm (hunk 6 above)
Comment (dklawren) [reply extensions/Voting/Extension.pm:903]: https://github.com/mozilla-bteam/bmo/blob/master/extensions/Voting/Extension.pm#L911

It happens when a bug is updated and triggers a voting change, we want the voting related comment to have the same timestamp as the other bug change(s).

File: Bugzilla/Report/Ping/Simple.pm (hunk 5 above)
Comment (dklawren) [reply Bugzilla/Report/Ping/Simple.pm:52]: Yes this is due to the newer version of JSON::Validator due to the new bmo-slim-perl image built with the latest CPAN dependencies. I can pin the version of JSON::Validator to an older version and build a new image. The reason for the new image in the first place was due to needing new version of the mysql perl bindings to support the move the 8.x.

File: Bugzilla/Bug.pm (hunk 2 above)
Comment (dklawren) [reply Bugzilla/Bug.pm:4183]: PostgreSQL uses double quotes instead and I do not want to knowingly break Pg. Using ->quote_identifier allows for any other changes needed in the future.

File: cpanfile.snapshot (hunk 1 above)
Comment (dklawren) [reply cpanfile.snapshot:25]: It is like a requirements.txt. The snapshot file is the list of the actual versions that were installed when all of the latest packages were pulled down. I will see if I can minimize the amount of changes.

**Comment by dklawren** [comment]
Sorry about the extra commits. I git rebased from main into this branch. Also I know why the tests are failing right now as I need to push the new image to docker hub and restart them.

**Comment by mitchhentges** [comment]
No worries, just ping me when this is ready for re-review :)

**Comment by dklawren** [comment]
Ugh it also added all of the files from the rebased commits to this PR. Hmm what did i do

**Comment by dklawren** [comment]
Ok. cleaned up with force push (sorry) and added a couple more places that needed to be updated.

File: Bugzilla/Report/Ping/Simple.pm (hunk 5 above)
Comment (mitchhentges) [reply Bugzilla/Report/Ping/Simple.pm:52]: Ah, that's fair.
Would you mind splitting the bumping of the new `bmo-slim-perl` image (plus fixing deprecated usages, such as `JSON::Validator` here) from the MSYQL8 changes?

That way, if there's risk of breakage, it may be easier to track down which change caused it (changed system + dependencies versus the signficant MySQL change).

File: Bugzilla/DB/Mysql.pm [hunk 7]
```diff
@@ -332,10 +332,17 @@ sub bz_setup_database {
   if ($self->utf8_charset eq 'utf8mb4') {
     my %global = map {@$_}
       @{$self->selectall_arrayref(q(SHOW GLOBAL VARIABLES LIKE 'innodb_%'))};
+
+    # Newer versions of MySQL the default value for innodb_file_format is Barracuda
```
Comment (mitchhentges) [inline Bugzilla/DB/Mysql.pm:336]:
```suggestion
    # In newer versions of MySQL the default value for innodb_file_format is Barracuda
```

Also, if you could specify a rough range (is "in MySQL 8+" correct?) that would help for readers of this comment a few years from now, when "newer version of MySQL" means something different :)

File: Bugzilla/Install/DB.pm [hunk 8]
```diff
@@ -1933,12 +1931,11 @@ sub _convert_groups_system_from_groupset {
 
     # Identify admin group.
     my ($admin_gid)
-      = $dbh->selectrow_array("SELECT id FROM groups WHERE name = 'admin'");
+      = $dbh->selectrow_array("SELECT id FROM `groups` WHERE name = 'admin'");
```
Comment (mitchhentges) [inline Bugzilla/Install/DB.pm:1934]: Let's do `$dbh->quote_identifier()` here

File: Bugzilla/Search.pm [hunk 9]
```diff
@@ -1435,8 +1435,9 @@ sub _translate_join {
   if ($extra_condition) {
     $extra_condition = " AND $extra_condition";
   }
-
-  my @join_sql = "$join JOIN $table AS $name"
+  my $sql_table
+    = $table eq 'groups' ? Bugzilla->dbh->quote_identifier($table) : $table;
```
Comment (mitchhentges) [inline Bugzilla/Search.pm:1439]: We could simplify this by unconditionally quote-ing it, though I'm happy either way.

File: userprefs.cgi [hunk 10]
```diff
@@ -533,17 +533,18 @@ sub DoPermissions {
...
-      . ") ORDER BY name");
+      . ') ORDER BY name');
   foreach my $group (@$groups) {
     my ($nam, $desc) = @$group;
     push(@has_bits, {"desc" => $desc, "name" => $nam});
   }
   $groups = $dbh->selectall_arrayref(
     'SELECT DISTINCT id, name, description
-                                          FROM groups
+                                          FROM `groups`
```
Comment (mitchhentges) [inline userprefs.cgi:547]: For consistency, let's `$dbh->quote_identifier()` here

**Review by mitchhentges — APPROVED** [review]
Looks good, nice work on the migration.
There's a couple missed `$dbh->quote_identifier()` calls that should probably be addressed, and from a commit-structure POV I think it'd be better to split the perl package (and docker) changes from the MySQL changes (though this part is more of a nit).

Send it 🚀

File: Bugzilla/Search.pm (hunk 9 above)
Comment (dklawren) [reply Bugzilla/Search.pm:1439]: I tried this in some cases the $table value is actually another string of SQL such as "foo JOIN ON (SELECT blah FROM blar)" So I can only quote if table is groups.

File: Bugzilla/Install/DB.pm (hunk 8 above)
Comment (dklawren) [reply Bugzilla/Install/DB.pm:1934]: Missed that one.
